Add regression tests for slow sentencepiece tokenizers.

PhilipMay opened the pull request:
This PR adds regression tests for slow sentencepiece tokenizers. These tests are needed for a refactoring in PR #11716 

## Strange findings
- s2t: `_convert_token_to_id` of `"<s>"` does not give 0

## ToDo
- <s>add test for `_convert_token_to_id`</s> - done, see `test_convert_token_and_id`
- <s>add test for `_convert_id_to_token`</s> - done, see `test_convert_token_and_id`
- <s>add test for `get_vocab`</s> - done
- <s>add test for `vocab_size`</s> - done
- <s>add test for `convert_tokens_to_string`</s> - done, see `test_sentencepiece_tokenize_and_convert_tokens_to_string` in `TokenizerTesterMixin` 
- <s>add test for pickle</s> - is tested in `test_pickle_subword_regularization_tokenizer`
- <s>manual review</s> - done
- <s>fix / add reformer integration test</s> - see https://github.com/huggingface/transformers/pull/11737#issuecomment-850769064 - done
- <s>add typing</s> - done
- <s>add docstrings</s> - done

PhilipMay commented:
rebased on master

PhilipMay commented:
This PR is ready for review please. @LysandreJik @sgugger 

The failing test is connected to #11731

sgugger reviewed (approved):
LGTM! Thanks a lot for adding those!

LysandreJik reviewed (approved):
Cool, thanks a lot for working on these tests! I think that these are already somewhat covered by the common tests, but they're fast and should help identify issues faster.

However, in order to make sure PR #11716 can be merged, I was mentioning integration tests, rather than regression/unit tests. For example the ALBERT integration test:

https://github.com/huggingface/transformers/blob/b8344a274fe13b390fa60c74b76117f5ea8144cb/tests/test_tokenization_albert.py#L108-L152

Those are particularly important when doing refactors that may affect the encoding/decoding aspect of tokenizers.

I think this is a bit of a larger work though, so we can post "Good first issues" for the SPM-based tokenizers in a first step so that the community may help.

PhilipMay commented:
> 
> 
> Cool, thanks a lot for working on these tests! I think that these are already somewhat covered by the common tests, but they're fast and should help identify issues faster.
> 
> However, in order to make sure PR #11716 can be merged, I was mentioning integration tests, rather than regression/unit tests. For example the ALBERT integration test:
> 
> https://github.com/huggingface/transformers/blob/b8344a274fe13b390fa60c74b76117f5ea8144cb/tests/test_tokenization_albert.py#L108-L152
> 
> Those are particularly important when doing refactors that may affect the encoding/decoding aspect of tokenizers.
> 
> I think this is a bit of a larger work though, so we can post "Good first issues" for the SPM-based tokenizers in a first step so that the community may help.

Ok @LysandreJik .
So I will extend the PR and add integration tests for the `_tokenizer` function like the one you linked above to all sentencepiece tokenizers.

Do you think the already written tests can stay as they are? What other steps are needed?

LysandreJik commented:
Hi @PhilipMay, thanks for offering to do it! Feel free to let us know if you would like us to offer some of these to the community, as it can be a bit of work to get every tokenizer tested.

Other than the integration tests, I don't think anything is needed.

Also, you might be interested in rebasing on the `master` branch - we've solved the issue regarding the `run_tests_torch` timing out yesterday so by rebasing you would have reliable CI feedback.

PhilipMay commented:
> Hi @PhilipMay, thanks for offering to do it! Feel free to let us know if you would like us to offer some of these to the community, as it can be a bit of work to get every tokenizer tested.

I was thinking to add integration tests for the tokeinzers that I want to refactor (the sentencepiece). And not foll all tokenizers.

**What about this:** In this PR I add integration tests for the **sentencepiece** tokeinzers only - a full list see here #11417 

After that has been merged I (or you) will open an issue asking for similar tests for all tokenizers.

@LysandreJik what do you think?

LysandreJik commented:
Yes, sentencepiece tokenizers only, definitely! But even so, that's quite a large number of tokenizers :)

PhilipMay commented:
> Also, you might be interested in rebasing on the master branch - we've solved the issue regarding the run_tests_torch timing out yesterday so by rebasing you would have reliable CI feedback.

Rebased on master - CI is green again. :-)

PhilipMay commented:
@LysandreJik I refactored the tokenizer integration test of albert;

https://github.com/German-NLP-Group/transformers/blob/1894bcc5d116d0107150f4659551e6e21111d736/tests/test_tokenization_albert.py#L127

By adding a util class to `TokenizerTesterMixin`
https://github.com/German-NLP-Group/transformers/blob/1894bcc5d116d0107150f4659551e6e21111d736/tests/test_tokenization_common.py#L186

Ans also added an integration test to Barthez:
https://github.com/German-NLP-Group/transformers/blob/1894bcc5d116d0107150f4659551e6e21111d736/tests/test_tokenization_barthez.py#L99

What do you think about this "pattern"? Should I dontinue in that direction and add the other tokenizers?

LysandreJik commented:
I like it, I find it very clean!

PhilipMay commented:
@LysandreJik the reformer tokenizer integration test somehow fails:

```text
Expected :Transformers (formerly known as pytorch-transformers and pytorch-pretrained-bert) provides general-purpose architectures (BERT, GPT-2, RoBERTa, XLM, DistilBert, XLNet...) for Natural Language Understanding (NLU) and Natural Language Generation (NLG) with over 32+ pretrained models in 100+ languages and deep interoperability between Jax, PyTorch and TensorFlow.
Actual   :Transformers (formerly known as pytorch-transformers and pytorch-pretrained-bert provides general-purpose architectures (BERT, GPT-, RoBERTa, LM, DistilBert, LNet... for Natural Language nderstanding (NL and Natural Language Generation (NLG with over  pretrained models in  languages and deep interoperability between ax, PyTorch and TensorFlow.
```

How does it come that it eats some characters? Any idea?

PhilipMay commented:
@LysandreJik while we are here - can we remove this or is it some kind of "open todo"?

https://github.com/huggingface/transformers/blob/8d171628fe84bdf92ee40b5375d7265278180f14/tests/test_tokenization_common.py#L178-L184

PhilipMay commented:
@LysandreJik @sgugger as discussed above the suggested integration tests are added to the sentencepiece tokenizers.

IMO this is done and ready for merge.

Please have a look at the strange behavior of the reformer tokenizer: https://github.com/huggingface/transformers/pull/11737#issuecomment-850769064

And this question: https://github.com/huggingface/transformers/pull/11737#issuecomment-850776366

LysandreJik commented:
Pinging @patrickvonplaten regarding the Reformer test.

Regarding https://github.com/huggingface/transformers/pull/11737#issuecomment-850776366 we can just remove this

LysandreJik commented on tests/test_tokenization_reformer.py:
```diff
@@ -328,3 +347,24 @@ def test_torch_encode_plus_sent_to_model(self):
+        # That is the reason why we use very simple texts here.
+        # Also see https://github.com/huggingface/transformers/pull/11737#issuecomment-850769064
+        sequences = [
+            "This is a very simple sentence." "The quick brown fox jumps over the lazy dog.",
```
Is this supposed to be two different sequences? It looks like it's a single sequence concatenated

PhilipMay replied:
Oops - good catch! I fixed it.

LysandreJik reviewed (approved):
This is fantastic, thanks a lot for working on this @PhilipMay!

PhilipMay commented:
> Regarding #11737 (comment) we can just remove this

Done.

patrickvonplaten commented on tests/test_tokenization_reformer.py:
```diff
@@ -328,3 +347,25 @@ def test_torch_encode_plus_sent_to_model(self):
+        expected_encoding = {'input_ids': [[108, 265, 24, 111, 4, 258, 156, 7, 51, 279, 58, 7, 76, 25, 69, 278], [140, 243, 264, 134, 17, 267, 77, 263, 22, 262, 297, 258, 304, 177, 279, 266, 14, 89, 13, 35, 261, 299, 272, 137, 275, 278]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}  # noqa: E ...
+        # fmt: on
+
+        # This tokenizer does not know some characters like ")".
```
Looks good to me!

PhilipMay commented:
Ok - so this should be ready to be merged so I can continue with #11716 ?

sgugger commented:
Thanks again for all your work!
